File: October LeetCoding Challenge/Week Four/Linked List Cycle II.py
class Solution(object):
    def detectCycle(self, head):
        if head == None: return None
        if head.next == None: return None
        
        first = head.next
        second = head.next.next
        
        while first != second:
            
            if first == None: return None
            if first.next == None: return None
            if second == None: return None
            if second.next == None: return None
            
            first = first.next
            second = second.next.next
            
        first = head
        
        while first != second:
            first = first.next
            second = second.next
        return first

File: Random/04-random-30/Linked List Cycle II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */

var detectCycle = function(head) {
    let slow = head;
    let fast = head;
    
    const detectIfLoop = () => {
        while(fast){
            if(fast.next && fast.next.next){
                slow = slow.next;
                fast = fast.next.next;
            }
            else fast = null;
            if(fast == slow) return true;
        }
        return false;
    }
    
    let loopExists = detectIfLoop();
    
    if(!loopExists) return null;
    else {
       let slow = head;
        while(slow != fast){
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
    
};
